README.md
# KineticReading
Integration of the Save Data Project, the Tech Dev Python Analytic Tools and the thrombophilia projects to automate analysis and file movement
  * GUI interface layer at KineticReading.py
  * CLI interface layer at kinetic_assay_cli.py
  * KineticAnalysis Class layer for analysis
  * Function Layer not necessary for Object Oriented Programming in KineticAssayTools

# Requirements
* Verbose error logging on GUI
* Automated GUI identification of labels, reducing the need to go into files themselves
* User can name labels whatever they want under fluor_cli.py/abs_cli.py layer
* User can name files whatever they want in whatever location they prefer
* User can analyze this at their desk and on their own computer easily
* Minimal IT help needed due to stable releases
* Tested at every layer in batch scripts and python files

## Versioning
Versions are tagged in git using the "Semantic Versioning 2.0" standard beginning with 0.1.0.
For details, please refer to: [http://semver.org/spec/v2.0.0.html](http://semver.org/spec/v2.0.0.html)

Version is manually updated using a text source file, but also provided automatically at build time by git itself, using it's `git describe` command.

## Summary
A version is of the form MAJOR.MINOR.PATCH, where each gets incremented for the following:

 * MAJOR, for incompatible API changes
 * MINOR, for functionality added in a backwards-compatible manner
 * PATCH, for backwards-compatible bug fixes

 The MAJOR patch will be zero for initial dev.

## Branching Model
Development is intended to follow the Git Flow branching model.

For details, please refer to: [http://nvie.com/posts/a-successful-git-branching-model/](http://nvie.com/posts/a-successful-git-branching-model/)

In summary, development takes place in feature branches (named anything except master, develop, release-\*, or hotfix-\*), then moved into release branches (release-\*), then tagged when released, and merged back into development and master branches.  For our purposes, build artifacts (elf, bin, hex, map files) are included in the release branches so that the build output is available once tagged.

## Install Anaconda 5.0.1 (64 bit), python 2.7.14
https://www.anaconda.com/download/
Anaconda>=5.0.1
python>=2.7.14
## Ensure relevant modules
relevant modules
matplotlib>=2.1.0
numpy>=1.13.3
scikit-learn>=0.19.1
statsmodels>=0.8.0
pyqt


releasenotes.txt
========================================
Release Notes
Automated Kinetic Analysis (aka KineticReading)
========================================

--------------------------------------
Release v0.2.5 (July 28, 2018)
--------------------------------------
New Features:
Prompts Logged Error if af/ff aren't in the same folder
Prompts Logged Error if users's supplied directories don't exist
Better logging, no need for all the junk time data
Cleaner UI, Adds assumptions to project for user to read
Handles saturated detectors, not just one but when all droplets are saturated and no slope can be calculated
Smart sizing of the droplet RFU figure to keep proportions right


Notes:
 *

Changes:
 *

Bug Fixes:
 * Handles saturated detectors, not just one but when all droplets are saturated and no slope can be calculated

Known Issues:
 *


--------------------------------------
Release v0.2.4 (July 13, 2018)
--------------------------------------
New Features:
Ability to measure a single droplet not just multiple

Notes:
 *

Changes:
 *

Bug Fixes:
 * (Critical Fix) Can now have a single droplet measured

Known Issues:
 *



--------------------------------------
Release v0.2.3 (June 04, 2018)
--------------------------------------
New Features:
GUI Layer
 * New organization of labels
 * Separation of abs/fluor buttons
 * Coloration schema to match Baebies colors
 * Automation of filling more label schemes
Kinetic Analysis CLI
 * start number/end number
 * filename output
 * normalization_factor
 * no integration of save data file
 Kinetic Tools
 * Create Size Dictionary used to check to make sure value pairs are all equivalent length
 * Used to check which drops have failed detections and which have mislabeled or uneven numbers of time/detection labels
 * GUI prompt of which drops had these issues

Notes:
 *

Changes:
 *

Bug Fixes:
 * Can now have different numbers of droplet detections for different drops

Known Issues:
 *


test_KineticReading.py
from KineticAssayTools import *
import os

def test_create_size_dictionary():
    dictionary = {'Hello 1': [3,4,5],'Droplet 1': [4,5,6,7], 'Background 1': [6,7,8,9], 'Hello 2':[5,6,7], 'Hello 3':[9,10,11,12]}
    output_dictionary = create_size_dictionary(dictionary)
    print output_dictionary
    assert output_dictionary['1'] == False
    assert output_dictionary['2'] == True
    assert output_dictionary['3'] == True
    print "Testing Creating dictionary sizing passes"

def test_return_key_labels():
    output,numbers = return_key_labels('OutputSaveFilew_label.csv', 'C:\Users\dhull\Documents\code_directory\KineticReading\\test\\testing')
    assert output == ['Drop_', 'Blank_']
    output, numbers = return_key_labels('OutputSaveFilewo_label.csv', 'C:\Users\dhull\Documents\code_directory\KineticReading\\test\\testing')
    assert output == ['Drop_', 'Blank_']
    output, numbers = return_key_labels('OutputSaveFilemislabeled.csv', 'C:\Users\dhull\Documents\code_directory\KineticReading\\test\\testing')
    assert len(output) == 3
    assert numbers == range(1,7)
    os.chdir('C:\Users\dhull\Documents\code_directory\KineticReading')
    print "Returning Key Labels Passes"


def test_get_sec():
    one_sec = get_sec(401000,400000)
    assert one_sec == 1.0
    twenty_sec = get_sec(420000,400000)
    assert twenty_sec == 20.0
    print "Getting Seconds Passes"


def test_name_data_file():

    '''# should rename file
    name_data_file('kinetic.log', os.getcwd())
    # should move file to new folder and rename flag as true
    name_data_file('kinetic.log', os.getcwd(), new_path_location=os.getcwd(), rename_flag=False)
    # should move file to a new folder with old name
    name_data_file('kinetic.log', os.getcwd(), new_path_location=os.getcwd(), rename_flag=True)
    '''
    pass

def test_calculate_OD():
    import numpy as np
    dark = np.array([0,0,0,0,0])
    ref = np.array([1,1,1,1,1])
    samp = np.array([[0.1,0.1,0.1,0.1,0.1],[0.01, 0.01, 0.01, 0.01, 0.01]])
    wavelengths = np.array([100, 200, 300, 400, 500])
    numpy_out = calculate_OD(dark, ref, samp, False, wavelengths)
    output = [[1,1,1,1,1],[2,2,2,2,2]]
    np.testing.assert_array_almost_equal(numpy_out,output)
    print "Calculate OD Passes"


def test_data_extraction():
    from KineticAnalysis import KineticAnalysis
    time_labels = ['Merge Drop ', 'Background Detect Time ', 'Droplet Detect Time ']
    fluor_labels = ['Blank_', 'Drop_']
    drop_numbers = range(1,8)
    KA_object = KineticAnalysis('AbsorbanceSaveFile.csv', 'OutputSaveFile.csv', 'C:\Users\dhull\Documents\code_directory\KineticReading\\test\sample_data_folder', drop_numbers, time_labels, fluor_labels)
    KA_object.data_extraction_ff()
    assert len(KA_object.fluor_dict.keys()) == 14
    for key in KA_object.fluor_dict.keys():
        assert len(KA_object.fluor_dict[key]) == 10
    assert KA_object.fluor_dict['Drop_1'] == [104.377, 103.049,	1000000,	102.699,	1000000,	101.884,	1000000,	100.309,	101.318,	100.154]
    print "Data Extraction Passes"

def test_get_timestamp_file():
    from KineticAnalysis import KineticAnalysis
    time_labels = ['Merge Drop ', 'Background Detect Time ', 'Droplet Detect Time ']
    fluor_labels = ['Blank_', 'Drop_']
    drop_numbers = range(1,8)
    KA_object = KineticAnalysis('AbsorbanceSaveFile.csv', 'OutputSaveFile.csv', 'C:\Users\dhull\Documents\code_directory\KineticReading\\test\sample_data_folder', drop_numbers, time_labels, fluor_labels)
    KA_object.get_timestamp_file('C:\Users\dhull\Documents\code_directory\KineticReading\\test\sample_data_folder\\results')
    assert KA_object.tf == '13-19-08.csv'
    assert KA_object.tf_loc == 'C:\Users\dhull\Documents\code_directory\KineticReading\\test\sample_data_folder\\results\\031418'
    print 'Get Timestamp File Passes'


def main():
    test_get_sec()
    test_name_data_file()
    test_return_key_labels()
    test_create_size_dictionary()
    test_calculate_OD()
    test_data_extraction()
    test_get_timestamp_file()

main()


KineticReading.py
import sys, os, csv, logging
from PyQt5.QtWidgets import QMainWindow, QApplication, QFileDialog
from PyQt5 import uic
from PyQt5.QtGui import QPixmap, QFont
from PyQt5 import QtCore

import sys, xlsxwriter
import numpy as np
sys.dont_write_bytecode = True
from KineticAnalysis import *
from KineticAssayTools import write_excel_file, name_data_file, return_key_labels, update_log

base_dir = os.getcwd()
Ui_MainWindow, QtBaseClass = uic.loadUiType("KineticReadingUI.ui")

class KRApp(QMainWindow):
    def __init__(self):
        super(KRApp, self).__init__()
        self.ui = Ui_MainWindow()
        self.ui.setupUi(self)
        self.clicked_run = 0
        self.ui.RunQpb.clicked.connect(self.run_experiments)
        self.ui.ResToDefQpb.clicked.connect(self.reset_defaults)
        self.ui.ResToLasExpQpb.clicked.connect(self.reset_recent_exp)
        self.ui.GetAbs.clicked.connect(self.getAbs)
        self.ui.GetFlu.clicked.connect(self.getFluor)
        self.ui.GetSavLoc.clicked.connect(self.getPath)
        self.ui_list = [self.ui.FluFilQte, self.ui.AbsFilQte, self.ui.AbsFilFluSavLocQte, self.ui.TimFilSavLocQte, self.ui.StaDroQte, self.ui.EndDroQte, self.ui.BlaLabQte, self.ui.DroLabQte,\
        self.ui.DarAbsLabQte, self.ui.BluAbsLabQte, self.ui.RefAbsLabQte, self.ui.MerDroLabQte, self.ui.BacDetLabQte, self.ui.DroDetLabQte, self.ui.Offset1, self.ui.Offset2, self.ui.SavQte, \
        self.ui.NorFacQte, self.ui.KinFilQte]

    def reset_defaults(self):
        """
        Grabs default csv which should be underneath temptext (hardcoded)
        repopulates UI with this information
        """
        os.chdir(base_dir)
        with open('defaults.csv') as csvfile:
            reader_object = csv.reader(csvfile)
            row_1 = reader_object.next()
            for index ,ui_element in enumerate(self.ui_list):
                ui_element.setText(row_1[index])

    def reset_recent_exp(self):
        """
        Grabs reset file and populates all ui Qte objects with previous
        experimental info
        """
        os.chdir(base_dir)
        with open('recent_exp.csv') as csvfile:
            reader_object = csv.reader(csvfile)
            row_1 = reader_object.next()
            for index ,ui_element in enumerate(self.ui_list):
                ui_element.setText(row_1[index])

    def getAbs(self):
        """
        Utilizes QFileDialog for finding and getting abs file
        """
        fname = QFileDialog.getOpenFileName(self, 'Open file',
         base_dir,"CSV files (*.csv)")
        split_by_path = fname[0].split('/')
        file = split_by_path[-1]
        split_by_path.pop()
        full_path = '\\'.join(split_by_path)
        self.ui.AbsFilQte.setText(file)
        self.ui.AbsFilFluSavLocQte.setText(full_path)
        all_labels, numbers = return_key_labels(file, full_path)
        list_labels = ','.join(all_labels)
        self.ui.DarAbsLabQte.setText(list_labels)
        self.ui.BluAbsLabQte.setText(list_labels)
        self.ui.RefAbsLabQte.setText(list_labels)

    def getFluor(self):
        """
        Utilizes QFileDialog to grab the fluor file
        """
        fname = QFileDialog.getOpenFileName(self, 'Open file',
         base_dir,"CSV files (*.csv)")
        split_by_path = fname[0].split('/')
        file = split_by_path[-1]
        split_by_path.pop()
        full_path = '\\'.join(split_by_path)
        self.ui.FluFilQte.setText(file)
        self.ui.AbsFilFluSavLocQte.setText(full_path)
        all_labels, numbers = return_key_labels(file, full_path)
        list_labels = ','.join(all_labels)
        self.ui.BlaLabQte.setText(list_labels)
        self.ui.DroLabQte.setText(list_labels)
        self.ui.StaDroQte.setText(str(min(numbers)))
        self.ui.EndDroQte.setText(str(max(numbers)))

    def getPath(self):
        """
        Gets path of the directory the user selects
        """
        fname = QFileDialog.getExistingDirectory(self, 'Select Directory', base_dir)
        self.ui.SavQte.setText(fname)

    def run_experiments(self):
        """
        Performs a number of tasks in order as described by the comments
        """
        os.chdir(base_dir)
        # writes data to the local csv file for feature "reset last experiment"
        # for next time if user wants to repopulate
        my_row = []
        for index ,ui_element in enumerate(self.ui_list):
            my_row.append(ui_element.toPlainText())
        with open('recent_exp.csv', 'wb') as csvfile:
            csvwriter_object = csv.writer(csvfile)
            csvwriter_object.writerow(my_row)
        # logs from parent directory
        logging.basicConfig(format='%(asctime)s %(message)s',\
        filename='kinetic.log',level=logging.DEBUG, datefmt= '%I:%M:%S %p')
        font = QFont()
        font.setPointSize(8)
        self.ui.LogLabQl.setFont(font)

        try:
            # populates relevant, frequently used variables for analysis from GUI
            drop_numbers = range(int(self.ui.StaDroQte.toPlainText()), int(self.ui.EndDroQte.toPlainText())+1)
            blank_drop_labels = [self.ui.BlaLabQte.toPlainText(), self.ui.DroLabQte.toPlainText()]
            time_stamp_events_list = [self.ui.MerDroLabQte.toPlainText(), self.ui.BacDetLabQte.toPlainText(), self.ui.DroDetLabQte.toPlainText()]
            absorbance_file_labels = [self.ui.DarAbsLabQte.toPlainText(), self.ui.BluAbsLabQte.toPlainText(), self.ui.RefAbsLabQte.toPlainText()]
            offset_bounds = [int(self.ui.Offset1.toPlainText()), int(self.ui.Offset2.toPlainText())]
            loop_labels = [blank_drop_labels, time_stamp_events_list, absorbance_file_labels]
            labels_w_spaces = [item for list_one in loop_labels for item in list_one if len(item.split(' ')) > 1]
            filename = self.ui.KinFilQte.toPlainText()
            normalization_factor = np.float64(self.ui.NorFacQte.toPlainText())

            # handles user incorrectly typing in information
            if not drop_numbers:
                logging.error('ERROR: Start Drop > End Drop')
                raise ValueError('ERROR: Start Drop > End Drop')
            try:
                directory_list = os.listdir(self.ui.AbsFilFluSavLocQte.toPlainText())
            except WindowsError:
                logging.error('ERROR: Could not find absorbance fluorescence file location')
                raise WindowsError('ERROR: Could not find absorbance fluorescence file location')
            if self.ui.AbsFilQte.toPlainText() not in directory_list:
                logging.debug('DEBUG: No abs file in supplied directory')
            if self.ui.FluFilQte.toPlainText() not in directory_list:
                logging.error('ERROR: Could not identify fluor file in supplied directory')
                raise WindowsError('Please check to make sure the path is correct for the fluor file')

            # Performs needed calculations
            KA_object = KineticAnalysis(self.ui.AbsFilQte.toPlainText(), self.ui.FluFilQte.toPlainText(), self.ui.AbsFilFluSavLocQte.toPlainText(), drop_numbers, time_stamp_events_list, blank_drop_labels)
            KA_object.get_timestamp_file(self.ui.TimFilSavLocQte.toPlainText())
            KA_object.data_extraction_log()
            KA_object.data_extraction_ff()
            if self.ui.AbsFilQte.toPlainText():
                KA_object.oil_blue_spectrum_analysis(absorbance_file_labels)
            KA_object.functional_assay_data_organization()
            KA_object.kinetic_analysis(offset_bounds, normalization_factor)
            KA_object.plot_drops_blanks_background()

            output_file = filename.split('.')[0] + '.xlsx'
            workbook_object = xlsxwriter.Workbook(output_file)
            KA_object.slopes['T 0 Slope Detection'] = str(offset_bounds[0])
            KA_object.slopes['T End Slope Detection'] = str(offset_bounds[1])

            # creates excel file and needed tabs
            write_excel_file(workbook_object, [KA_object.drop_dictionary], 'Times and RFU Counts')
            write_excel_file(workbook_object, [KA_object.slopes], 'Slope data')
            if self.ui.AbsFilQte.toPlainText():
                KA_object.adjusted_slopes['T 0 Slope Detection'] = str(offset_bounds[0])
                KA_object.adjusted_slopes['T End Slope Detection'] = str(offset_bounds[1])
                write_excel_file(workbook_object, [KA_object.adjusted_slopes], 'Adjusted Slopes')
                write_excel_file(workbook_object, [np.array([KA_object.path_length]), 'OilBlue.png'], 'OilBlueImage')
            write_excel_file(workbook_object, ['Droplet_RFU_Figure.png'], 'Droplet RFU Figure')
            workbook_object.close()

            # moves data
            os.chdir(base_dir)
            if self.ui.SavQte.toPlainText():
                name_data_file(KA_object.tf, KA_object.tf_loc,new_path_location=self.ui.SavQte.toPlainText(), rename_flag=False)
                name_data_file(self.ui.FluFilQte.toPlainText(), self.ui.AbsFilFluSavLocQte.toPlainText(), new_path_location=self.ui.SavQte.toPlainText(), rename_flag=True)
                if self.ui.AbsFilQte.toPlainText():
                    name_data_file(self.ui.AbsFilQte.toPlainText(), self.ui.AbsFilFluSavLocQte.toPlainText(), new_path_location=self.ui.SavQte.toPlainText(), rename_flag=True)
                name_data_file(output_file, self.ui.AbsFilFluSavLocQte.toPlainText(), new_path_location=self.ui.SavQte.toPlainText(), rename_flag=True)

            font = QFont()
            font.setPointSize(24)
            self.ui.LogLabQl.setFont(font)
            self.ui.LogLabQl.setText('Success!')
        except:
            logging.error('ERROR: AN ERROR HAS OCCURRED')
            update_log(self.ui, base_dir, os.getcwd())

        # truncates the log with each new run
        os.chdir(base_dir)
        with open('kinetic.log', 'w'):
            pass

from PyQt5.QtWidgets import QApplication
app = QApplication(sys.argv)
window = KRApp()
window.show()
sys.exit(app.exec_())


KineticAnalysis.py
'''
Developer: Daniel Hull dhull@baebies.com
Plese reach out with any questions or points of improvement
Copyright Baebies 2018
See Readme for additional details
'''

class KineticAnalysis:
    def __init__(self, af_file, ff_file, af_ff_save_location, drop_numbers, base_time_labels, base_fluor_labels):
        # import relevant modules
        import numpy as np
        import csv, os, logging
        from statsmodels.nonparametric.smoothers_lowess import lowess
        import matplotlib.pyplot as plt
        global plt, np, csv, os, lowess

        self.ff_file = ff_file
        self.af_file = af_file
        self.af_ff_save_location = af_ff_save_location
        self.drop_numbers = drop_numbers
        self.time_labels = [label+str(i) for label in base_time_labels for i in drop_numbers]
        self.assay_labels = [label+str(i) for label in base_fluor_labels for i in drop_numbers]
        self.label_bases = base_time_labels + base_fluor_labels
        logging.info('Kinetic Analysis Begun')

    def get_timestamp_file(self, results_directory):
        """
        Adapted from tech_dev_python_analytics_tools.py
        Assumes sorted data folders based on date

        Inputs:
        results_directory - (str) that houses the log file
        """

        import os, logging
        from time import strftime, localtime
        try:
            os.chdir(results_directory)
        except WindowsError:
            logging.info('ERROR: Failed to Grab Time File, directory doesn''t exist')
        current_directory_contents = os.listdir(os.getcwd())
        most_recent_date = current_directory_contents[-1]
        # changes directory, grabs most recent file
        os.chdir(results_directory+'\\'+most_recent_date)
        files = os.listdir(os.getcwd())
        results_file_csv = files[-1]
        self.tf = results_file_csv
        self.tf_loc = os.getcwd()
        logging.info('Successfully got timestamp file')
        logging.info('Timestamp file is ' + str(results_file_csv))

    def data_extraction_ff(self):
        """
        Maps fluorescent data from known columns of labels and values
        Assumes you are looking down a column, a fair assumption
        Raises error if your fluorimeter stopped working at any point

        Inputs: self.af_ff_save_location
        Outputs: self.fluor_dict
        """
        import csv, os, time, logging
        from KineticAssayTools import create_size_dictionary
        os.chdir(self.af_ff_save_location)
        label_col = 0
        data_col = 8
        self.fluor_dict = {key:[] for key in self.assay_labels}
        with open(self.ff_file, 'rb') as csvfile:
            myread = csv.reader(csvfile)
            for row in myread:
                if row[label_col] in self.fluor_dict.keys():
                    self.fluor_dict[row[label_col]].append(float(row[data_col]))
        #handles failed detections
        error_check_dictionary = create_size_dictionary(self.fluor_dict)
        failed_drops = {key:val for key, val in error_check_dictionary.iteritems() if val is False}
        if not not failed_drops:
            logging.error('ERROR: Fluor Drop det != to blank det at drops ' + str(failed_drops.keys()))
            raise ValueError('ERROR: Failed Detection or uneven blank/drop combinations at Drops ' + str(failed_drops.keys()))

        #notifies user successful completion
        logging.info('Successfully extracted ff data')

    def data_extraction_log(self):
        """
        Maps fluorescent data from a csv file to a dictionary with
        Assumes you are looking down a column, a fair assumption for STS and fluorimeter firmware outputs

        Adds to self:
        log_dict: a key-value pair of labels and all corresponding data of interest, will convert to float, must be numerical data
        """
        import csv, os, logging
        label_col = 0
        data_col = 3
        self.log_dict = {key:[] for key in self.time_labels}
        with open(self.tf, 'rb') as csvfile:
            myread = csv.reader(csvfile)
            for row in myread:
                if row[label_col] in self.log_dict.keys():
                    self.log_dict[row[label_col]].append(float(row[data_col]))
        logging.info('Extracted time data')

    def functional_assay_data_organization(self):
        """
        Performs all relevant calculations including background subtraction and time conversion to seconds differentials
        For all - there is a single merge time that needs to be subtracted
        Handles two frequent user errors

        Inputs: self log_dict, fluor_dict

        Outputs: drop_dictionary
        """
        from KineticAssayTools import get_sec, create_size_dictionary
        import logging

        #handles cases where there are dual experiments, or no time file mapped or not enough or no time labels
        #filters out merge keys because those will be length one, it's always the first key in the list
        without_merge = list(filter(lambda x: self.label_bases[0] not in x, self.log_dict.keys()))
        with_merge = list(filter(lambda x: self.label_bases[0] in x, self.log_dict.keys()))
        for key in with_merge:
            if len(self.log_dict[key])!=1:
                logging.error('ERROR: Some merges are not found in the time file')
                raise ValueError('ERROR: Some merges are not found in the time file' )
        temp_dict = {k: self.log_dict[k] for k in without_merge}
        temp_dict.update(self.fluor_dict)
        error_check_dictionary = create_size_dictionary(temp_dict)
        failed_drops = {key:val for key, val in error_check_dictionary.iteritems() if val is False}
        if not not failed_drops:
            logging.info('ERROR: Diff # of det labels and actual det at Drops' + str(failed_drops.keys()))
            raise ValueError('ERROR: Different Number of Detection labels and actual detections at Drop ' + str(failed_drops.keys()))

        self.drop_dictionary = {}
        for i in self.drop_numbers:
            try:
                self.drop_dictionary['Droplet_Detect_Time_From_Merge_'+str(i)]  = [get_sec(val, self.log_dict[self.label_bases[0] + str(i)][0]) for val in self.log_dict[self.label_bases[2] + str(i)]]
                self.drop_dictionary['Blank_Detect_Time_From_Merge_'+str(i)]  = [get_sec(val, self.log_dict[self.label_bases[0] + str(i)][0]) for val in self.log_dict[self.label_bases[2] + str(i)]]
                self.drop_dictionary['DropSub_'+str(i)] = [a_i - b_i for a_i, b_i in zip(self.fluor_dict[self.label_bases[4]+str(i)], self.fluor_dict[self.label_bases[3]+str(i)])]
            except KeyError:
                logging.error('ERROR: Drop # in time file or fluor file do not line up with the labels on UI')
                raise KeyError('')
        self.drop_dictionary.update(self.fluor_dict)
        logging.info('successfully blank subtracted, calculated det times relative to merge')

    def oil_blue_spectrum_analysis(self, dark_ref_blue_list):
        """
        Oil Blue Spectrum Analysis, adaptation from the TechDevPythonAnalyticsTools
        Adapts key_word_search, path_length_estimation, & basic_plot

        Inputs:
        csv_sort_names: names of the reference, dark, and  (list, len == 3)

        Self Outputs:
        path_length - the path length of the cartridge
        """

        from KineticAssayTools import calculate_OD
        import numpy as np
        import logging

        # reads in absorbance csv
        self.csv_lol = []
        with open(self.af_file, 'rb') as csvfile:
            myread = csv.reader(csvfile)
            for row in myread:
                self.csv_lol.append(row)
        self.csv_lol = np.array(self.csv_lol)
        # identifies the locations of the key words of dark, ref, & blue
        dark_ref_blue_dict = {key : np.where(self.csv_lol == key) for key in dark_ref_blue_list}

        sample_row = np.array(self.csv_lol[0])
        # grabs the wavelength outputs from the
        spectrophotometer_wavelengths = []
        wavelengths_begin = [index+1 for index, val in enumerate(sample_row) if val == 'wavelengths']
        spectrophotometer_wavelengths = sample_row[wavelengths_begin[0]:wavelengths_begin[0]+1024]
        self.spectrophotometer_wavelengths = spectrophotometer_wavelengths.astype(np.float)

        # extracts a matrix of the counts and puts them in each dictionary, assumes the counts begin after the second firmware output of the word with spectrum
        counts_begin = [index for index, val in enumerate(sample_row) if 'spectrum' in val.split('_')]
        self.counts_begin = counts_begin[1]+1

        dark_counts_lol = [self.csv_lol[i][self.counts_begin:] for i in dark_ref_blue_dict[dark_ref_blue_list[0]][0]]
        ref_counts_lol = [self.csv_lol[i][self.counts_begin:] for i in dark_ref_blue_dict[dark_ref_blue_list[2]][0]]
        blue_counts_lol = [self.csv_lol[i][self.counts_begin:] for i in dark_ref_blue_dict[dark_ref_blue_list[1]][0]]

        # converts to numpy array
        dark_counts_lol = np.array(dark_counts_lol).astype(np.int)
        ref_counts_lol = np.array(ref_counts_lol).astype(np.int)
        blue_counts = np.array(blue_counts_lol).astype(np.int)

        # takes average of ref and dark
        self.dark_counts = np.mean(dark_counts_lol, axis=0)
        self.ref_counts = np.mean(ref_counts_lol, axis=0)
        # calculates OD of blue oil absorbance, multiple reads made
        OD = calculate_OD(self.dark_counts, self.ref_counts, blue_counts, True, self.spectrophotometer_wavelengths)
        #self.blue_counts = lowess(OD, self.spectrophotometer_wavelengths, frac=0.02, return_sorted=False)
        self.path_length = 10000*max(OD[647:668]-np.median(OD[813:833]))/1.51

        # plots oil blue spectrum
        plt.clf()
        plt.plot(self.spectrophotometer_wavelengths, OD)
        plt.xlim([500, 700])
        plt.xlabel('wavelengths (nm)')
        plt.ylim([0,0.1])
        plt.ylabel('OD')
        path_length = '%s' %float('%g6' %self.path_length)
        plt.text(600, 0.075, 'pathlength is ' + path_length)
        plt.title('Blue Oil Absorbance')
        plt.savefig('OilBlue.png')
        logging.info('calculated OD and created Oil Blue plot')

    def kinetic_analysis(self, offset_bounds, normalization_factor):
        """
        Kinetic analysis by running a linear regression on the RFUs of assays over time
        Assumptions:
            each set is row wise not column wise, so whatever you want analyzed, make it a row

        Self:
        slopes - numpy array of slopes for each data set organized by rows
        """

        from sklearn import linear_model
        import logging
        import numpy as np
        regr = linear_model.LinearRegression()
        self.slopes = {}
        self.adjusted_slopes = {}
        self.intercept = {}
        for drop_number in self.drop_numbers:
            if not offset_bounds:
                background_sub_lol = np.array(self.drop_dictionary['DropSub_'+str(drop_number)][:])
                drop_time_lol = np.array(self.drop_dictionary['Droplet_Detect_Time_From_Merge_'+str(drop_number)][:])
            else:
                background_sub_lol = np.array(self.drop_dictionary['DropSub_'+str(drop_number)][offset_bounds[0]-1:offset_bounds[1]])
                drop_time_lol = np.array(self.drop_dictionary['Droplet_Detect_Time_From_Merge_'+str(drop_number)][offset_bounds[0]-1:offset_bounds[1]])
            elements_to_delete = [col for col, val in enumerate(background_sub_lol) if (abs(val) > 999000) or (val ==0)]
            background_sub_lol = np.delete(background_sub_lol, elements_to_delete, None)
            drop_time_lol =  np.delete(drop_time_lol, elements_to_delete, None)
            if background_sub_lol.size != 0:
                self.slopes['Drop Number '+ str(drop_number)] = np.array([])
                self.intercept['Drop Number '+ str(drop_number)] = np.array([])
                regr.fit(drop_time_lol.reshape(len(drop_time_lol),1), background_sub_lol.reshape(len(drop_time_lol),1))
                self.slopes['Drop Number '+ str(drop_number)] =  [regr.coef_[0][0]]
                self.slopes['Inst Factor Slope of Drop Number '+ str(drop_number)] = [normalization_factor*regr.coef_[0][0]]
                self.intercept['Drop Number '+ str(drop_number)] = [regr.intercept_[0]]
                if self.af_file:
                    self.adjusted_slopes['Drop Number '+ str(drop_number)] = [(self.slopes['Drop Number '+ str(drop_number)][0])*(10000.0/self.path_length)]
                    self.adjusted_slopes['Normalized by OD & Inst Slope of Drop Number '+ str(drop_number)] =  [normalization_factor*(self.slopes['Drop Number '+ str(drop_number)][0])*(10000.0/self.path_length)]
            else:
                pass
        logging.info('completed kinetic analysis')

    def plot_drops_blanks_background(self):
        """
        Adapted from single plex script

        Inputs:

        Returns: None
        """

        import matplotlib.pyplot as plt
        import numpy as np

        if len(self.drop_numbers) > 1:
            fig, axs = plt.subplots(1,len(self.drop_numbers))
            axs = axs.reshape(-1)
        else:
            fig, axs = plt.subplots(1,len(self.drop_numbers)+1)
            axs = axs.reshape(-1)
        for pos, drop_number in enumerate(self.drop_numbers):
            try:
                self.slopes['Drop Number '+ str(drop_number)] = np.around(self.slopes['Drop Number '+ str(drop_number)], decimals=3)
                self.intercept['Drop Number '+ str(drop_number)] = np.around(self.intercept['Drop Number '+ str(drop_number)], decimals=3)
                my_text = 'y='+str(self.slopes['Drop Number '+ str(drop_number)][0])+'x+'+str(self.intercept['Drop Number '+ str(drop_number)][0])
                filtered_output_drops = list(filter(lambda x: (x>0 and x!=1000000), self.drop_dictionary['Drop_' + str(drop_number)]))
                filtered_output_blanks = list(filter(lambda x: (x>0 and x!=1000000), self.drop_dictionary['Blank_' + str(drop_number)]))
                filtered_output_back_sub =  list(filter(lambda x: (x>-300 and x<3000), self.drop_dictionary['DropSub_'+str(drop_number)]))
                filtered_list = filtered_output_drops + filtered_output_blanks + filtered_output_back_sub
                (y_max, y_min) = (max(filtered_list), min(filtered_list))
                axs[pos].scatter(self.drop_dictionary['Blank_Detect_Time_From_Merge_'+str(drop_number)], self.drop_dictionary['Blank_'+str(drop_number)], c='b')
                axs[pos].scatter(self.drop_dictionary['Droplet_Detect_Time_From_Merge_'+str(drop_number)], self.drop_dictionary['Drop_' + str(drop_number)], c='g')
                axs[pos].scatter(self.drop_dictionary['Droplet_Detect_Time_From_Merge_'+str(drop_number)], self.drop_dictionary['DropSub_'+str(drop_number)], c='r')
                axs[pos].set_xlabel('seconds post merge')
                axs[pos].set_ylabel('RFUs')
                axs[pos].set_title('Drop ' + str(drop_number))
                axs[pos].set_ylim(top = y_max+10, bottom=y_min-10)
                axs[pos].text(0.5, 0.3, my_text, horizontalalignment='center', verticalalignment='center', transform = axs[pos].transAxes)
            except:
                pass
        length = 6*len(self.drop_numbers)
        fig.set_size_inches(length, 7.5)
        plt.savefig('Droplet_RFU_Figure.png')
